[Hyun] step-2 구현 : 회원의 수정, 질문글쓰기 구현

ghkdgus29 opened the pull request:
요구사항들이 늘어나면서 리포지토리에서의 작업들이 복잡해지는 느낌을 받았습니다.  
조금 더 복잡해진다면 서비스계층을 만드는 것을 고려해보아야 할 것 같습니다.

또한 회원 수정 기능을 넣으면서, 회원 등록 폼과 회원 수정 폼을 만들었습니다.
등록 폼과 수정 폼은 다른 방식으로 데이터를 처리하기 때문에 UserUpdate DTO 를 만들까도 생각해보았습니다.

일단은 update 시에 int id 와 user 객체 (기존 비밀번호가 들어있는 객체입니다.), String newPassword 를 매개변수로 넘겨 검증 및 수정을 수행하도록 하였습니다.

그러나 @ModelAttribute User updateUser 로 가져온 user 객체가 기존 바뀌지 않은 비밀번호를 가지고 있다는 것이 직관적이지 않아 따로 DTO를 만드는 것이 더 깔끔할 것 같기도 합니다. 이에 대해서 고민중입니다.

HTML Form 에서는 POST 와 GET 전송만 가능하다고 알고 있었는데 일종의 꼼수를 사용하여 PUT 이나 DELETE 전송을 할 수 있다는 사실을 알았습니다. 
가급적 동사를 사용하지 않는 것이 좋다고 생각하여 POST : users/{id}/update 대신 PUT : users/{id} 를 사용하였습니다.
URL 네이밍은 항상 고민되는 것 같습니다.

ghkdgus29 committed:
fix: index.html 과 user/form.html 을 정적 페이지로 처리

ghkdgus29 committed:
feat: 타임리프 레이아웃을 사용하여 중복 제거

ghkdgus29 committed:
fix: 유저 목록이 순서대로 찍히도록 타임리프를 사용하여 수정

1, 2, 3 ... 이런 식으로 출력된다.

ghkdgus29 committed:
feat: 게시글을 담는 Article 클래스와 게시글을 저장하는 리포지토리를 생성하고, 기본적인 저장과 조회기능을 구현

저장 -> save
조회 -> findById, findAll

ghkdgus29 committed:
feat: 질문하기 POST 요청 기능 구현

ArticleController 의 question 메서드가 POST HTML Form 을 받아 article 객체를 생성하고 저장
그 후 / 으로 redirect

ghkdgus29 committed:
fix: 게시글을 동적으로 렌더링하기 위해, index.html 을 static 에서 templates 로 이동

ghkdgus29 committed:
feat: 홈으로 redirect 시 article 저장소에서 모든 글 목록들을 불러와 출력

타임리프를 사용하여 글 목록들을 출력

ghkdgus29 committed:
feat: article 목록 초기데이터 저장

눈으로 확인하기 위해 2개의 글을 미리 저장

ghkdgus29 committed:
feat: 글 목록에 댓글수와 글 작성 시간 출력

Article 객체에 replyCount 필드와 createDate 필드 추가

ghkdgus29 committed:
feat: 게시글 상세보기 기능 구현

ghkdgus29 committed:
feat: User 클래스에 id 정수를 저장

ghkdgus29 committed:
feat: 회원 정보 수정 화면 출력 기능 구현

ghkdgus29 committed:
feat: 회원정보 수정 기능 구현

hidden type을 사용하여 PUT 으로 전송

ghkdgus29 committed:
feat: 회원정보 수정 시, 비밀번호를 검증 후 회원정보를 수정 가능하도록 구현

만약 비밀번호가 틀릴경우 예외를 던진다.
새 비밀번호를 따로 입력받아 변경
MemoryUserRepository 클래스내에 update 메서드가 수정을 담당

honux77 commented:
현의 고민들이 바람직하고 좋은 것 같습니다. 필요하다면 DTO나 서비스 모두 만들 수 있겠죠?

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,62 @@
+    @GetMapping("/articles/{index}")
+    public String showDetailedArticle(@PathVariable int index, Model model) {
+        log.info("{}", index);
+        Article article = repository.findById(index);
```
게시글이 존재하지 않으면 어떻게 되는 걸까요?

ghkdgus29 replied:
적절한 예외처리를 하지 않아 ArrayIndexOutOfBoundsException 이 터지게 되네요!!
해당 내용은 id 검증을 리포지토리에 넣어주어 수정했습니다!! 

피드백 감사합니다!

ghkdgus29 committed:
feat: 존재하지 않는 게시글 검색 시 예외를 던지도록 검증 기능 구현

MemoryArticleRepository 의 findById 메서드 내에서 검증을 먼저 수행한 뒤 Article을 찾아서 반환한다.
존재하지 않는 Article 검색 시엔 예외를 던진다.
